### PR TITLE
Webpack

### PR DESCRIPTION
**Pull request by kayone**
Switching to WebPack has a few advantages.

First and the main one is getting rid of requirejs.
The 2nd one is we can switch from the AMD module format to commonjs format,

```
define(
    [
        'marionette'
    ], function (Marionette) {

        return Marionette.CompositeView.extend({
            template: 'Series/Index/EmptyTemplate'
        });
    });
```

becomes

```
var marionette = require('marionette');
module.exports = marionette.CompositeView.extend({
    template : 'Series/Index/EmptyTemplate'
});;
```

specially helps with modules that have tons of dependencies, 

it also has support for source maps so we can debug a deployed build like we do in dev.

and generally a much more flexible loader, It'll allow us to do better caching (you can break the file into components (settings, system, series, history etc)) and they get their own hash breaker/file automatically.

### CODE REVIEW

**Comment by kayone** [comment]
Here is the diffrence between r.js and webpack

**two main advantages right off the bat, no more version cache breaking, everything is done using hashes, so backend changes won't unnecessarily invalidate the cache.
Every file gets a cache breaker (even main.js), so it's much safer. 
We don't have to load r.js anymore. save 20K there.
And as I mentioned before we can even break the 2MB main js into separate packs so app is loaded much fast and cache is invalidated less often. so potential savings are huge**

### R.JS
image


### Webpack
image

File: gulp/webpack.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+var gulp = require('gulp');
+
+var run = require('gulp-run');
+
+
+gulp.task('webpack', function (cb) {
+  run('webpack -d').exec()
```
Comment (Taloth) [inline gulp/webpack.js]: Shouldn't you use something like gulp-webpack here?

File: package.json [hunk 2]
```diff
@@ -26,10 +26,9 @@
...
     "gulp-print": "1.1.0",
     "gulp-replace": "0.5.0",
+    "gulp-run": "1.6.6",
     "gulp-wrap": "0.5.0",
-    "gulp-wrap-amd": "0.3.1",
     "jshint-stylish": "1.0.0",
-    "requirejs": "2.1.15",
     "run-sequence": "1.0.2",
     "streamqueue": "0.1.1"
   }
```
Comment (Taloth) [inline package.json]: Missing webpack/gulp-webpack here, I had to install webpack manually from npm to make it compile.
If you use gulp-webpack you can probably remove the gulp-run again.

File: src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs [hunk 3]
```diff
@@ -27,6 +27,7 @@ public override bool CanHandle(string resourceUrl)
         {
             return resourceUrl.StartsWith("/Content") ||
                    resourceUrl.EndsWith(".js") ||
+                   resourceUrl.EndsWith(".map") ||
```
Comment (Taloth) [inline src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs]: .map is for the sourcemap right, or are we talking about a different kind of map? Isn't a sourcemap embedded in the .js better?

File: src/UI/.idea/encodings.xml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="Encoding" useUTFGuessing="true" native2AsciiForPropertiesFiles="false" />
-</project>
-
+  <component name="Encoding" useUTFGuessing="true" native2AsciiForPropertiesFiles="false">
+    <file url="file://$PROJECT_DIR$/System/Logs/Files/LogFileModel.js" charset="UTF-8" />
```
Comment (Taloth) [inline src/UI/.idea/encodings.xml]: Adding this encoding probably wasn't intentional.

File: src/UI/Shims/jquery.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+var jquery = require('../JsLibraries/jquery');
+
+require('../Instrumentation/StringFormat');
+
+window.$= jquery;
```
Comment (Taloth) [inline src/UI/Shims/jquery.js]: Nitpick: space before =

File: webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      'backbone'                : 'Shims/backbone',
+      'moment'                  : 'JsLibraries/moment',
+      'filesize'                : 'JsLibraries/filesize',
+      'handlebars'              : 'Shared/Shims/handlebars',
+      'handlebars.helpers'      : 'JsLibraries/handlebars.helpers',
+      'bootstrap'               : 'JsLibraries/bootstrap',
+      'backbone.deepmodel'      : 'JsLibraries/backbone.deep.model',
+      'backbone.pageable'       : 'JsLibraries/backbone.pageable',
+      'backbone-pageable'       : 'JsLibraries/backbone.pageable',
+      'backbone-pageable'       : 'JsLibraries/backbone.pageable',
```
Comment (Taloth) [inline webpack.config.js]: Duplicate with '-', and another one with  '.', how about picking one?

**Comment by Taloth** [comment]
I'm not experienced at all at this, so I just skimmed over it. 
Overall it looks like an excellent idea, afaik we don't have too many outstanding UI changes. But if we do this, we have to do this in one go, rebasing this stuff would be absolute hell.

File: gulp/webpack.js (hunk 1 above)
Comment (kayone) [inline gulp/webpack.js]: the watcher in webpack is much more efficient that if we used gulp and re-ran the whole thing.

File: src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs (hunk 3 above)
Comment (kayone) [inline src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs]: embedding the source map would kind of defeat it's purpose of trying to minimize the amount of data that needs to be sent to the client.

File: webpack.config.js (hunk 6 above)
Comment (kayone) [inline webpack.config.js]: so there was a reason for this, depending on modules (3rd party) that depend on backbone.pagable and how they reference it. sometime they require it by `-` sometime by `.`

we can re-evalute this, but this is straight from the require js config.

**Comment by kayone** [comment]
Yeah, specially the change that I haven't pushed to this branch, it touches EVERY single file, in a big way.

File: gulp/webpack.js (hunk 1 above)
Comment (Taloth) [inline gulp/webpack.js]: from what I can see gulp-webpack has a 'watch: true' option.

File: src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs (hunk 3 above)
Comment (Taloth) [inline src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs]: You would only include the sourcemap in debug builds right?

Comment (kayone) [inline src/NzbDrone.Api/Frontend/Mappers/StaticResourceMapper.cs]: source maps are in a different file, so we can have them included in the release builds as well, they don't get downloaded unless dev tools is open.

File: webpack.config.js (hunk 6 above)
Comment (Taloth) [inline webpack.config.js]: I can understand the . and -, but - is list twice.
I understand it's straight from the require file, doesn't mean the duplicate shouldn't be fixed.

**Comment by Taloth** [comment]
One problem: I tried it out, but if I change for example Router.js, webpack doesn't refresh the main.js, could be something wrong with my system, so please verify and lemme know.

File: webpack.config.js (hunk 6 above)
Comment (kayone) [inline webpack.config.js]: yeah, I fixed the duplicate, I was just explaining the `-` and `.`

**Comment by kayone** [comment]
no, it's not you, It used to work, it doesn't work for me here either. :(

**Pull request merged**